src/daemon/rpmostreed-os.c

```diff
@@ -77,6 +82,109 @@ sysroot_changed (RpmostreedSysroot *sysroot,
+                     GDBusMethodInvocation  *invocation)
+{
+  RpmostreedOS *self = RPMOSTREED_OS (interface);
+  const gchar *method_name = g_dbus_method_invocation_get_method_name (invocation);
+  const gchar *sender = g_dbus_method_invocation_get_sender (invocation);
+  GVariant *parameters = g_dbus_method_invocation_get_parameters (invocation);
+  const gchar *action = NULL;
+  gboolean authorized = FALSE;
+
+  if (g_strcmp0 (method_name, "GetDeploymentsRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "GetCachedDeployRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "DownloadDeployRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "GetCachedUpdateRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "DownloadUpdateRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "GetCachedRebaseRpmDiff") == 0 ||
+      g_strcmp0 (method_name, "DownloadRebaseRpmDiff") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.repo-refresh";
+    }
+  else if (g_strcmp0 (method_name, "Rebase") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.rebase";
+    }
+  else if (g_strcmp0 (method_name, "Deploy") == 0 ||
+           g_strcmp0 (method_name, "Upgrade") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.upgrade";
+    }
+  else if (g_strcmp0 (method_name, "SetInitramfsState") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.set-initramfs-state";
+    }
+  else if (g_strcmp0 (method_name, "Cleanup") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.cleanup";
+    }
+  else if (g_strcmp0 (method_name, "Rollback") == 0 ||
+           g_strcmp0 (method_name, "ClearRollbackTarget") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.rollback";
+    }
+  else if (g_strcmp0 (method_name, "PkgChange") == 0)
+    {
+      action = "org.projectatomic.rpmostree1.install-uninstall-packages";
+    }
+  else if (g_strcmp0 (method_name, "UpdateDeployment") == 0)
+    {
+      g_autoptr(GVariant) modifiers = g_variant_get_child_value (parameters, 0);
+      g_auto(GVariantDict) dict;
+      g_variant_dict_init (&dict, modifiers);
+      const char *refspec =
+        vardict_lookup_ptr (&dict, "set-refspec", "&s");
+      g_autofree char **install_pkgs =
+        vardict_lookup_ptr (&dict, "install-packages", "^a&s");
+      g_autofree char **uninstall_pkgs =
+        vardict_lookup_ptr (&dict, "uninstall-packages", "^a&s");
+      g_autoptr(GVariant) install_local_pkgs =
+        g_variant_dict_lookup_value (&dict, "install-local-packages",
+                                     G_VARIANT_TYPE("ah"));
+
+      if (refspec != NULL)
+        action = "org.projectatomic.rpmostree1.rebase";
+      else if (install_pkgs != NULL || uninstall_pkgs != NULL)
+        action = "org.projectatomic.rpmostree1.install-uninstall-packages";
+      else if (install_local_pkgs != NULL)
+        action = "org.projectatomic.rpmostree1.install-local-packages";
+      else
+        action = "org.projectatomic.rpmostree1.upgrade";
```

Comment:
Hmm, yeah these checks are messy indeed since multiple actions can happen. Is it possible here to accumulate actions and then check that the user is authorized for all of them? Otherwise it's just going to be messy how to determine the "most privileged action", which itself might depend on e.g. the site policy.